### DISCUSSION TITLE
libstdc++.so.6: version `GLIBCXX_3.4.29' not found

### DISCUSSION THREAD

**Discussion by shizuku** [pybind/pybind11#3453] [Q&A]
I run this in pybind11
```
mkdir build
cd build
cmake ..
make check -j 4
```
got ERROR:
```
......
Traceback (most recent call last):
  File "/home/shizuku/code/python/pybind11/tests/test_cmake_build/subdirectory_function/../test.py", line 4, in <module>
    import test_cmake_build
ImportError: /home/shizuku/.conda/envs/py3.9/bin/../lib/libstdc++.so.6: version `GLIBCXX_3.4.29' not found (required by /home/shizuku/code/python/pybind11/build/tests/test_cmake_build/subdirectory_function/test_cmake_build.cpython-39-x86_64-linux-gnu.so)
......
```
I run `cmake_example` got ERROR:
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ImportError: /home/shizuku/.conda/envs/py3.9/bin/../lib/libstdc++.so.6: version `GLIBCXX_3.4.29' not found (required by /home/shizuku/.conda/envs/py3.9/lib/python3.9/site-packages/cmake_example.cpython-39-x86_64-linux-gnu.so)
```
my OS: Arch Linux
python: 3.9.7 (Miniconda)

**Comment by lucasjinreal** [comment]
@shizuku For anyone might got this error, a simple way to fix it is by copy or system libstdc++ to conda according folder.

in my case:

```
rm /home/xx/anaconda3/bin/../lib/libstdc++.so.6
cp /usr/lib/x86_64-linux-gnu/libstdc++.so.6.0.29 /home/xx/anaconda3/bin/../lib
ln -s /home/xx/anaconda3/bin/../lib/libstdc++.so.6.0.29 /home/xx/anaconda3/bin/../lib/libstdc++.so.6
```

then

```
 strings /home/xx/anaconda3/bin/../lib/libstdc++.so.6 | grep GLIBCXX_3.4.29                                     1 main!?
GLIBCXX_3.4.29
```

**Reply by kenfehling** [reply]
Thanks, I actually made a symbolic link instead of copying. Worked well!

**Reply by loriayu** [reply]
我的服务器中没有6.0.29版本，从网上下载后拷贝进去进行ln也可以使用，谢谢。

**Reply by Jet-Shawn** [reply]
This is the only way for me to solve the problem. Thanks a lot!!!

**Comment by Ssakura-go** [comment]
I met the same problem and successfully fixed it by installing the opencv / scikit-image in the right version. In addition, maybe the version of pip should be taken into account.

**Reply by Lan0508** [reply]
reinstalling scikit-image works👍

**Comment by modu001** [comment]
My ERROR was
`/lib/x86_64-linux-gnu/libstdc++.so.6: version 'GLIBCXX_3.4.29' not found`
In this case, I was able to solve the problem by copying the anaconda lib.
1. `strings /home/xx/anaconda3/envs/xx/lib/libstdc++.so.6 | grep GLIBCXX` to confirm that GLIBCXX_3.4.29 exists in one of your environments.
2. `rm /lib/x86_64-linux-gnu/libstdc++.so.6`
3. `cp /home/xx/anaconda3/envs/xx/lib/libstdc++.so.6 /lib/x86_64-linux-gnu`

**Reply by chfritz** [reply]
This undoes exactly what a virtual environment like anaconda is meant to achieve: *not* having to replace system libraries in order to satisfy dependencies.

**Reply by robofar** [reply]
@chfritz I have same issue as @modu001 . I git cloned repo and install it as package in my conda environment. But why does package .so file looks for this libstdc++.so.6 system-wide, why it does not look at the same file but in conda directory? I grep it and it contains `GLIBCXX_3.4.29`, but system-wide .so file does not contain it.

**Reply by rezajahadi** [reply]
Thank you! my problem was extacly same as you and it solved.

**Reply by HKAFITGlitter** [reply]
what if I do not have permission to copy file to /lib/x86_64-linux-gnu?

**Reply by rezajahadi** [reply]
I used sudo in order to have permission.

**Reply by luccachiang** [reply]
@robofar Hi, I have the same question. Why doesn't it search from the conda environment? Did you figure out the answer? I do not think modifying the file under `/lib` is a safe way since many users share one machine, though it does work around.

**Reply by justin4ai** [reply]
@HKAFITGlitter Hi, I'm in the exactly same path. Would you please let me know if you solved the problem? (no permission)

**Reply by vmahash-tataaig** [reply]
@luccachiang @justin4ai @robofar It is possible that the linker didn't find the `libstdc++.so.6` inside the conda environment directory, and thus defaulted to the system library. In my case it was because the one I had was named `libstdc++.so.6.0.34`, failing the 'file exists' check. This can be fixed by creating a symlink yourself, (as also discussed in https://github.com/pytorch/pytorch/issues/157709)

`cd /home/xx/anaconda3/envs/xx/lib/`
`ln -s libstdc++.so.6.0.34 libstdc++.so.6`

It is also possible that the environment variable `LD_LIBRARY_PATH` is missing the environment lib directory. It was the case on my Sagemaker instance, but the linking wasn't affected, so I am not sure if this one is relevant or not.

Ofc, this ought to be managed during instantiation of the environment or installation of relevant packages, but I am not sure how to trigger such behaviour manually.

**Comment by yash051** [comment]
I simply delete 'libstdc++.so.6' symbolic link file from 'miniconda3/envs/env_name/lib/' and thats it program run perfectly.

**Reply by Ruhrozz** [reply]
really helped!

**Reply by Ruhrozz** [reply]
I also had an error with Jupyter after that:
`ImportError: cannot import name 'AstRenderer' from 'mistune.renderers'`

but it was solved with:
`pip install mistune --upgrade`

I hope it will help somebody

**Reply by 1037861070** [reply]
good

**Reply by ZZQ987** [reply]
Thank you!

**Reply by Taylor-zym** [reply]
good, works for me

**Comment by ZenMoore** [comment]
I have a suggestion for all who encountered this issue:
Please first check the message after `required by`.
In my case, this version of GLIBCXX_3.4.29 is required by `xxx/miniconda3/envs/internlp/lib/python3.10/site-packages/pandas/_libs/window/aggregations.cpython-310-x86_64-linux-gnu.so`, which is not the same case as the one the blogger encountered.
As a result, I don't need to change the soft links.
All I need to do is just `pip install --upgrade pandas`.

**Reply by sysuqm** [reply]
You are right. It works.

**Reply by hibaozhong** [reply]
It works.

**Reply by mickelliu** [reply]
It works.

**Reply by aymuos15** [reply]
This worked for me.

**Reply by anki-code** [reply]
Solved by `pip install -U --force-reinstall scipy` in my case. Thanks!

**Reply by Imatgay** [reply]
@anki-code's solution worked for me too!

**Reply by yydsun** [reply]
有用！

**Reply by mrvgao** [reply]
WOW!!! You are a hero!!! It works!!!

> Solved by `pip install -U --force-reinstall scipy` in my case. Thanks!

**Reply by coconut0713** [reply]
> Solved by `pip install -U --force-reinstall scipy` in my case. Thanks!

Perfect!!! Thanks!!!!!

**Reply by shizhediao** [reply]
You saved my day!

**Reply by maiducbinh** [reply]
This worked for me!!! Tks <3.

**Reply by MyhooSalsa** [reply]
Solved by pip install -U --force-reinstall scipy in my case. Thanks!

**Reply by SagiPolaczek** [reply]
> Solved by `pip install -U --force-reinstall scipy` in my case. Thanks!

Solved mine too. Thanks :)

**Reply by vantienpham** [reply]
In my case, it related to PIL:
```
ImportError: /usr/lib/x86_64-linux-gnu/libstdc++.so.6: version `GLIBCXX_3.4.29' not found (required by /home/van-tien.pham/anaconda3/lib/python3.10/site-packages/PIL/../../.././libLerc.so.4)
```
Solved by:
```
pip uninstall pillow
pip install pillow
```
Thank you.

**Reply by feature-space-move** [reply]
God, you save me!

**Comment by nungsorb** [comment]
In my case, solved by 
`export LD_LIBRARY_PATH=/home/user/miniconda3/lib:$LD_LIBRARY_PATH
`
But ensure that LD_LIBRARY_PATH includes this lib directory so that the system uses the correct version of the libstdc++ library.

**Reply by yqkqknct** [reply]
This works, if you are running `mujoco_py` this may be relevant! (I installed osmesa to conda like https://github.com/ethz-asl/reinmav-gym/issues/35#issuecomment-1222946797)

**Comment by jinghaoliu** [comment]
step1: echo 'export LD_LIBRARY_PATH=/home/conda/envs/XXX/lib:$LD_LIBRARY_PATH' >> ~/.bashrc

step2: source ~/.bashrc

**Reply by Petrichor361** [reply]
Great! This works, because i have no admin permission.

**Reply by AlirezaSalehy** [reply]
That is wonderful! I couldn't believe it would be that simple. Many thanks!

**Reply by dveytia** [reply]
This was definitely the best solution -- THANK YOU!!!

**Reply by raidenMEI10** [reply]
Thanks a lot, that's the most fabulous way I have found in the github or csdn.

**Reply by pomu-A** [reply]
Thanks a lot!

**Comment by Cupcc** [comment]
1. find the higher version of `libstdc++.so.*`
  - `find / libstdc++.so.*`
2. check whether there is a saftisfied `GLIBCXX` version
  - `strings /to/path/libstdc++.so.6.0.29 | grep GLIBCXX`
2. copy the higher viersion file to the position where program cann't find
4. remove current softlink
  - `rm /to/path/libcstdc++.so.6`
5. add the new softlink with the higher version
  - `ln -s /to/path/libstdc++.so.6.0.29 /to/path/libstdc++.so.6`

**Comment by paulpacaud** [comment]
I solved this problem using:
conda install -c conda-forge libstdcxx-ng

**Comment by pomu-A** [comment]
For me, It solved by:
module load GCC/12.3.0

**Comment by liuyixin-louis** [comment]
resolve with module load GCC/13.3.1 or conda install libgcc-ng # For Linux
